ruby/lib/endpoints/types.rb

```diff
@@ -0,0 +1,633 @@
+          end
+        end
+
+        # Invokes a command on the resource. Commands are expected to be under the
+        # "commands/" sub-resource.
+        def _cmd(command, data = nil, params = nil, api_version = 1)
+          _post('commands/' + command, ApiCommand, false, data, params, api_version)
+        end
+
+        # Retrieves an ApiConfig list from the given relative path.
+        def _get_config(rel_path, view, api_version = 1)
+          _require_min_api_version(api_version)
+          params = view && { 'view' => view } || nil
+          resp = @_resource_root.get(_path + '/' + rel_path, params)
+          json_to_config(resp, view == 'full')
+        end
+
+        def _update_config(rel_path, config, api_version = 1)
+          _require_min_api_version(api_version)
+          resp = @_resource_root.put(_path + '/' + rel_path, nil, config_to_json(config))
+          json_to_config(resp, false)
+        end
+
+        def _delete(rel_path, ret_type, ret_is_list = false, params = nil, api_version = 1)
+          _call(:delete, rel_path, ret_type, ret_is_list, nil, params, api_version)
+        end
+
+        def _get(rel_path, ret_type, ret_is_list = false, params = nil, api_version = 1)
+          _call(:get, rel_path, ret_type, ret_is_list, nil, params, api_version)
+        end
+
+        def _post(rel_path, ret_type, ret_is_list = false, data = nil, params = nil, api_version = 1)
+          _call(:post, rel_path, ret_type, ret_is_list, data, params, api_version)
+        end
+
+        def _put(rel_path, ret_type, ret_is_list = false, data = nil, params = nil, api_version = 1)
+          _call(:put, rel_path, ret_type, ret_is_list, data, params, api_version)
+        end
+
+        def _call(method_name, rel_path, ret_type, ret_is_list = false, data = nil, params = nil, api_version = 1)
+          path = _path
+          path += '/' + rel_path if rel_path
+          call_resource(@_resource_root.method(method_name), path, ret_type, ret_is_list, data, params, api_version)
+        end
+      end
+
+      # A list of some api object
+      class ApiList < BaseApiObject
+        LIST_KEY = 'items'.freeze
+
+        def initialize(objects, resource_root = nil, *args)
+          super(resource_root, args)
+          instance_variable_set('@objects', objects)
+        end
+
+        def to_str
+          "<ApiList>(#{@objects.length}): [#{@objects.map(&:to_str).join(', ')}]"
+        end
+
+        def to_json_dict(preserve_ro = false)
+          ret = super
+          attr = Attr.new
+          res = []
+          @objects.each do |x|
+            res << attr.attr_to_json(x, preserve_ro)
+          end
+          ret[LIST_KEY] = res
+          ret
+        end
+
+        def length
+          @objects.length
+        end
+
+        def each(&block)
+          @objects.each(&block)
+        end
+
+        def [](i)
+          @objects[i]
+        end
```

Comment:
Got me here. The `[]` means what when invoking? Does this return a specific object from the list, like an array? `foo[4]`?